### PR TITLE
[SPARK-20373][SQL][SS] Batch queries with 'Dataset/DataFrame.withWatermark()` does not execute

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

Any Dataset/DataFrame batch query with the operation `withWatermark` does not execute because the batch planner does not have any rule to explicitly handle the EventTimeWatermark logical plan. 
The right solution is to simply remove the plan node, as the watermark should not affect any batch query in any way.

Changes: 
- In this PR, we add a new rule `EliminateEventTimeWatermark` to check if we need to ignore the event time watermark. We will ignore watermark in any batch query. 

Depends upon:
- [SPARK-20672](https://issues.apache.org/jira/browse/SPARK-20672). We can not add this rule into analyzer directly, because streaming query will be copied to `triggerLogicalPlan ` in every trigger, and the rule will be applied to `triggerLogicalPlan` mistakenly. 

Others:
- A typo fix in example.

## How was this patch tested?

add new unit test.

### CODE REVIEW

**Comment by uncleGen** [comment]
cc @zsxwing and @tdas

**Comment by uncleGen** [comment]
retest this please.

**Comment by uncleGen** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -2457,6 +2457,19 @@ object CleanupAliases extends Rule[LogicalPlan] {
 }
 
 /**
+ * Ignore event time watermark in batch query, which is only supported in Structured Streaming.
+ * TODO: add this rule into analyzer rule list.
+ */
+object CheckEventTimeWatermark extends Rule[LogicalPlan] {
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Could you rename it to `EliminateEventTimeWatermark` and add it into `Analyzer.batches` as one of `Substitution` rules?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -2457,6 +2457,19 @@ object CleanupAliases extends Rule[LogicalPlan] {
...
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case EventTimeWatermark(_, _, child) if !child.isStreaming =>
+      logWarning("EventTime watermark is only supported in Structured Streaming but found " +
```
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: this warning log is not needed. Could you just update `withWatermark` doc and says 

```
Using `withWatermark` on a non-streaming [[Dataset]] is no-op`.
```

**Comment by felixcheung** [comment]
Please rebase to pick up fix for R tests

**Comment by uncleGen** [comment]
Depends upon: [SPARK-20672](https://issues.apache.org/jira/browse/SPARK-20672)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (uncleGen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: @zsxwing This pr does some prepare work before we add `EliminateEventTimeWatermark ` into `Analyzer.batches`. Could you please take a review?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (uncleGen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: got

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (zsxwing) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I see. The current approach is good to me then. Could you rename it, please?

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
@uncleGen Thanks! LGTM. Merging to master and 2.2.

**Pull request closed without merging**